<a href="https://colab.research.google.com/github/luizkrawiec/estatistics/blob/main/estatistica_udemy.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Dados Enem 2019**

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

Exploração inicial dos Dados

In [3]:
dados = pd.read_csv('/content/microdados_enem_2019_sp.csv', sep=';', encoding='iso-8859-1')

dados.head()

NU_INSCRICAO  CO_MUNICIPIO_RESIDENCIA NO_MUNICIPIO_RESIDENCIA  \
0     1,90E+11                  3546801            Santa Isabel   
1     1,90E+11                  3538204             Pinhalzinho   
2     1,90E+11                  3550308               São Paulo   
3     1,90E+11                  3548708   São Bernardo do Campo   
4     1,90E+11                  3549409    São Joaquim da Barra   

   CO_UF_RESIDENCIA SG_UF_RESIDENCIA  NU_IDADE TP_SEXO  TP_ESTADO_CIVIL  \
0                35               SP        22       M                1   
1                35               SP        19       M                1   
2                35               SP        17       F                1   
3                35               SP        19       M                1   
4                35               SP        37       M                2   

   TP_COR_RACA  TP_NACIONALIDADE  ...  NU_NOTA_MT TP_LINGUA  \
0            3                 1  ...      6002.0         0   
1            3                 1  ...         NaN         1   
2            3                 1  ...      7318.0         0   
3            3                 1  ...      3932.0         1   
4            1                 1  ...         NaN         1   

   TP_STATUS_REDACAO NU_NOTA_COMP1  NU_NOTA_COMP2  NU_NOTA_COMP3  \
0                1.0         160.0          200.0          180.0   
1                NaN           NaN            NaN            NaN   
2                1.0         160.0          120.0          200.0   
3                1.0          80.0          160.0          120.0   
4                NaN           NaN            NaN            NaN   

   NU_NOTA_COMP4  NU_NOTA_COMP5  NU_NOTA_REDACAO  Q025  
0          200.0          200.0            940.0     B  
1            NaN            NaN              NaN     A  
2          200.0          200.0            880.0     A  
3          100.0          100.0            560.0     B  
4            NaN            NaN              NaN     B  

[5 rows x 44 columns]

In [4]:
dados.shape

(813772, 44)

O número total de inscritos em São Paulo foi de 813772, sendo 15,97% de todo o Brasil que foi de 5095308 inscritos.

In [5]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 813772 entries, 0 to 813771
Data columns (total 44 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   NU_INSCRICAO             813772 non-null  object 
 1   CO_MUNICIPIO_RESIDENCIA  813772 non-null  int64  
 2   NO_MUNICIPIO_RESIDENCIA  813772 non-null  object 
 3   CO_UF_RESIDENCIA         813772 non-null  int64  
 4   SG_UF_RESIDENCIA         813772 non-null  object 
 5   NU_IDADE                 813772 non-null  int64  
 6   TP_SEXO                  813772 non-null  object 
 7   TP_ESTADO_CIVIL          813772 non-null  int64  
 8   TP_COR_RACA              813772 non-null  int64  
 9   TP_NACIONALIDADE         813772 non-null  int64  
 10  CO_MUNICIPIO_NASCIMENTO  794063 non-null  float64
 11  NO_MUNICIPIO_NASCIMENTO  794063 non-null  object 
 12  CO_UF_NASCIMENTO         794063 non-null  float64
 13  SG_UF_NASCIMENTO         794063 non-null  object 
 14  TP_S

# **Limpeza de Dados**

In [6]:
dados1 = dados.drop(columns=['CO_MUNICIPIO_RESIDENCIA'])

In [7]:
dados1.head()

NU_INSCRICAO NO_MUNICIPIO_RESIDENCIA  CO_UF_RESIDENCIA SG_UF_RESIDENCIA  \
0     1,90E+11            Santa Isabel                35               SP   
1     1,90E+11             Pinhalzinho                35               SP   
2     1,90E+11               São Paulo                35               SP   
3     1,90E+11   São Bernardo do Campo                35               SP   
4     1,90E+11    São Joaquim da Barra                35               SP   

   NU_IDADE TP_SEXO  TP_ESTADO_CIVIL  TP_COR_RACA  TP_NACIONALIDADE  \
0        22       M                1            3                 1   
1        19       M                1            3                 1   
2        17       F                1            3                 1   
3        19       M                1            3                 1   
4        37       M                2            1                 1   

   CO_MUNICIPIO_NASCIMENTO  ... NU_NOTA_MT  TP_LINGUA TP_STATUS_REDACAO  \
0                1506500.0  ...     6002.0          0               1.0   
1                1503804.0  ...        NaN          1               NaN   
2                1501402.0  ...     7318.0          0               1.0   
3                3548708.0  ...     3932.0          1               1.0   
4                3549409.0  ...        NaN          1               NaN   

   NU_NOTA_COMP1  NU_NOTA_COMP2  NU_NOTA_COMP3  NU_NOTA_COMP4  NU_NOTA_COMP5  \
0          160.0          200.0          180.0          200.0          200.0   
1            NaN            NaN            NaN            NaN            NaN   
2          160.0          120.0          200.0          200.0          200.0   
3           80.0          160.0          120.0          100.0          100.0   
4            NaN            NaN            NaN            NaN            NaN   

   NU_NOTA_REDACAO  Q025  
0            940.0     B  
1              NaN     A  
2            880.0     A  
3            560.0     B  
4              NaN     B  

[5 rows x 43 columns]

In [8]:
dados1 = dados1.drop(columns=['CO_UF_RESIDENCIA','SG_UF_RESIDENCIA','CO_MUNICIPIO_NASCIMENTO'], )

In [9]:
dados1 = dados1.drop(columns=['NO_MUNICIPIO_NASCIMENTO','CO_UF_NASCIMENTO','SG_UF_NASCIMENTO',
                              'TP_ANO_CONCLUIU','TP_ENSINO','CO_MUNICIPIO_ESC','CO_UF_ESC','SG_UF_ESC'],)

In [10]:
dados1 = dados1.drop(columns=['TP_DEPENDENCIA_ADM_ESC','TP_LOCALIZACAO_ESC','TP_SIT_FUNC_ESC'],)

In [11]:
dados1.shape

(813772, 29)

Enriquecendo Dados

In [18]:
dados1.loc[:,'NU_NOTA_CN'] /= 10
dados1.loc[:,'NU_NOTA_CH'] /= 10
dados1.loc[:,'NU_NOTA_LC'] /= 10
dados1.loc[:,'NU_NOTA_MT'] /= 10

In [19]:
dados1.head()

NU_INSCRICAO NO_MUNICIPIO_RESIDENCIA  NU_IDADE TP_SEXO  TP_ESTADO_CIVIL  \
0     1,90E+11            Santa Isabel        22       M                1   
1     1,90E+11             Pinhalzinho        19       M                1   
2     1,90E+11               São Paulo        17       F                1   
3     1,90E+11   São Bernardo do Campo        19       M                1   
4     1,90E+11    São Joaquim da Barra        37       M                2   

   TP_COR_RACA  TP_NACIONALIDADE  TP_ST_CONCLUSAO  TP_ESCOLA  IN_TREINEIRO  \
0            3                 1                1          1             0   
1            3                 1                1          1             0   
2            3                 1                2          3             0   
3            3                 1                2          2             0   
4            1                 1                1          1             0   

   ...  NU_NOTA_MT TP_LINGUA  TP_STATUS_REDACAO  NU_NOTA_COMP1  NU_NOTA_COMP2  \
0  ...       600.2         0                1.0          160.0          200.0   
1  ...         NaN         1                NaN            NaN            NaN   
2  ...       731.8         0                1.0          160.0          120.0   
3  ...       393.2         1                1.0           80.0          160.0   
4  ...         NaN         1                NaN            NaN            NaN   

   NU_NOTA_COMP3  NU_NOTA_COMP4  NU_NOTA_COMP5  NU_NOTA_REDACAO  Q025  
0          180.0          200.0          200.0            940.0     B  
1            NaN            NaN            NaN              NaN     A  
2          200.0          200.0          200.0            880.0     A  
3          120.0          100.0          100.0            560.0     B  
4            NaN            NaN            NaN              NaN     B  

[5 rows x 29 columns]

Renomeando colunas

In [20]:
dados1 = dados1.rename(columns={'NU_NOTA_REDACAO': 'NOTA_REDACAO'})

In [21]:
dados1 = dados1.rename(columns={'NU_NOTA_CN': 'NOTA_CN','NU_NOTA_CH': 'NOTA_CH',
                                'NU_NOTA_LC': 'NOTA_LC','NU_NOTA_MT': 'NOTA_MT'})

In [22]:
dados1 = dados1.rename(columns={'NU_NOTA_COMP1': 'COMP1','NU_NOTA_COMP2': 'COMP2',
                                'NU_NOTA_COMP3': 'COMP3','NU_NOTA_COMP4': 'COMP4',
                                'NU_NOTA_COMP5':'COMP5'})

In [23]:
dados1 = dados1.rename(columns={'NU_IDADE': 'IDADE','TP_SEXO': 'SEXO',
                                'TP_COR_RACA': 'RACA','Q025': 'INTERNET',
                                'TP_ESCOLA':'ESCOLA'})

Renomeando registros

In [24]:
dados1["RACA"] = dados1["RACA"].replace({0:"nao_declarado", 1:"branca", 2:"preta", 3:"parda", 4:"amarela", 5:"indigena"})

In [25]:
dados1["TP_LINGUA"] = dados1["TP_LINGUA"].replace({0:"Inglês", 1:"Espanhol"})

In [26]:
dados1["ESCOLA"] = dados1["ESCOLA"].replace({1:"não_respondeu", 2:"pública", 3:"privada", 4:"exterior"})

In [27]:
dados1["INTERNET"] = dados1["INTERNET"].replace({"A":"Não", "B":"Sim"})

In [33]:
dados1.head()

NU_INSCRICAO NO_MUNICIPIO_RESIDENCIA  IDADE SEXO  TP_ESTADO_CIVIL    RACA  \
0     1,90E+11            Santa Isabel     22    M                1   parda   
1     1,90E+11             Pinhalzinho     19    M                1   parda   
2     1,90E+11               São Paulo     17    F                1   parda   
3     1,90E+11   São Bernardo do Campo     19    M                1   parda   
4     1,90E+11    São Joaquim da Barra     37    M                2  branca   

   TP_NACIONALIDADE  TP_ST_CONCLUSAO         ESCOLA  IN_TREINEIRO  ...  \
0                 1                1  não_respondeu             0  ...   
1                 1                1  não_respondeu             0  ...   
2                 1                2        privada             0  ...   
3                 1                2        pública             0  ...   
4                 1                1  não_respondeu             0  ...   

   NOTA_MT TP_LINGUA  TP_STATUS_REDACAO  COMP1  COMP2  COMP3  COMP4  COMP5  \
0    600.2    Inglês                1.0  160.0  200.0  180.0  200.0  200.0   
1      NaN  Espanhol                NaN    NaN    NaN    NaN    NaN    NaN   
2    731.8    Inglês                1.0  160.0  120.0  200.0  200.0  200.0   
3    393.2  Espanhol                1.0   80.0  160.0  120.0  100.0  100.0   
4      NaN  Espanhol                NaN    NaN    NaN    NaN    NaN    NaN   

   NOTA_REDACAO  INTERNET  
0         940.0       Sim  
1           NaN       Não  
2         880.0       Não  
3         560.0       Sim  
4           NaN       Sim  

[5 rows x 29 columns]

## Análise Exploratória e Transformação dos dados

**Análise das idades**

In [28]:
dados1['IDADE'].value_counts()


IDADE
17    219313
18    147548
19     81236
16     59316
20     53009
       ...  
86         1
4          1
3          1
91         1
2          1
Name: count, Length: 80, dtype: int64

In [29]:
dados1['IDADE'].value_counts().sort_index()

IDADE
2     1
3     1
4     1
5     2
6     2
     ..
80    2
81    3
82    2
86    1
91    1
Name: count, Length: 80, dtype: int64

In [30]:
# Para ver todas as linhas, primeiro transforme num Dataframe:
idade = pd.DataFrame(dados1['IDADE'].value_counts().sort_index())

In [31]:
# Configura o Pandas para observar 80 linhas
pd.set_option('display.max_rows', 80)

In [32]:
idade.head(80)

count
IDADE        
2           1
3           1
4           1
5           2
6           2
7           3
8           6
12          6
13         29
14        355
15      11583
16      59316
17     219313
18     147548
19      81236
20      53009
21      37242
22      29489
23      23531
24      18851
25      15103
26      12465
27      10367
28       8986
29       7650
30       6967
31       6331
32       5685
33       5175
34       4795
35       4340
36       4109
37       4105
38       3767
39       3299
40       3058
41       2638
42       2462
43       2307
44       2069
45       1939
46       1718
47       1578
48       1496
49       1304
50       1167
51       1014
52        933
53        841
54        727
55        652
56        530
57        497
58        388
59        325
60        254
61        236
62        176
63        144
64        139
65        115
66         81
67         72
68         53
69         43
70         40
71         25
72         17
73         16
74         13
75         10
76          8
77          3
78          3
79          4
80          2
81          3
82          2
86          1
91          1

Alguns valores estranhos e curiosos: idades menores que 12 anos e maiores de 80 anos.

In [34]:
menores_12 = dados1.query('IDADE < 12')['NO_MUNICIPIO_RESIDENCIA'].value_counts()

In [35]:
# Inscrições com idade abaixo de 12 anos serão excluídos devido a alta probabilidade de erro.
menores_12

NO_MUNICIPIO_RESIDENCIA
São Paulo                 5
Tambaú                    1
Ribeirão Preto            1
Bálsamo                   1
São Bernardo do Campo     1
Embu das Artes            1
Queiroz                   1
Santana de Parnaíba       1
Vargem Grande Paulista    1
Auriflama                 1
Araçatuba                 1
Campinas                  1
Name: count, dtype: int64

In [36]:
dados_maiores_11 = dados1.loc[dados1.IDADE > 11]

In [37]:
dados_maiores_11.head()

NU_INSCRICAO NO_MUNICIPIO_RESIDENCIA  IDADE SEXO  TP_ESTADO_CIVIL    RACA  \
0     1,90E+11            Santa Isabel     22    M                1   parda   
1     1,90E+11             Pinhalzinho     19    M                1   parda   
2     1,90E+11               São Paulo     17    F                1   parda   
3     1,90E+11   São Bernardo do Campo     19    M                1   parda   
4     1,90E+11    São Joaquim da Barra     37    M                2  branca   

   TP_NACIONALIDADE  TP_ST_CONCLUSAO         ESCOLA  IN_TREINEIRO  ...  \
0                 1                1  não_respondeu             0  ...   
1                 1                1  não_respondeu             0  ...   
2                 1                2        privada             0  ...   
3                 1                2        pública             0  ...   
4                 1                1  não_respondeu             0  ...   

   NOTA_MT TP_LINGUA  TP_STATUS_REDACAO  COMP1  COMP2  COMP3  COMP4  COMP5  \
0    600.2    Inglês                1.0  160.0  200.0  180.0  200.0  200.0   
1      NaN  Espanhol                NaN    NaN    NaN    NaN    NaN    NaN   
2    731.8    Inglês                1.0  160.0  120.0  200.0  200.0  200.0   
3    393.2  Espanhol                1.0   80.0  160.0  120.0  100.0  100.0   
4      NaN  Espanhol                NaN    NaN    NaN    NaN    NaN    NaN   

   NOTA_REDACAO  INTERNET  
0         940.0       Sim  
1           NaN       Não  
2         880.0       Não  
3         560.0       Sim  
4           NaN       Sim  

[5 rows x 29 columns]

In [38]:
dados_maiores_11['IDADE'].value_counts().sort_index()

IDADE
12         6
13        29
14       355
15     11583
16     59316
17    219313
18    147548
19     81236
20     53009
21     37242
22     29489
23     23531
24     18851
25     15103
26     12465
27     10367
28      8986
29      7650
30      6967
31      6331
32      5685
33      5175
34      4795
35      4340
36      4109
37      4105
38      3767
39      3299
40      3058
41      2638
42      2462
43      2307
44      2069
45      1939
46      1718
47      1578
48      1496
49      1304
50      1167
51      1014
52       933
53       841
54       727
55       652
56       530
57       497
58       388
59       325
60       254
61       236
62       176
63       144
64       139
65       115
66        81
67        72
68        53
69        43
70        40
71        25
72        17
73        16
74        13
75        10
76         8
77         3
78         3
79         4
80         2
81         3
82         2
86         1
91         1
Name: count, dtype: int64

**Treineiros**

In [39]:
dados_maiores_11.query('IN_TREINEIRO == 1')['IN_TREINEIRO'].value_counts()

IN_TREINEIRO
1    93988
Name: count, dtype: int64

In [40]:
treineiros = dados_maiores_11.loc[dados_maiores_11.IN_TREINEIRO == 1]

In [41]:
treineiros.head()

NU_INSCRICAO NO_MUNICIPIO_RESIDENCIA  IDADE SEXO  TP_ESTADO_CIVIL    RACA  \
20      1,90E+11                  Santos     15    F                1   parda   
21      1,90E+11                Panorama     16    F                1   parda   
106     1,90E+11               São Paulo     16    F                1  branca   
109     1,90E+11        Águas de Lindóia     16    M                1  branca   
159     1,90E+11        Águas de Lindóia     15    M                1  branca   

     TP_NACIONALIDADE  TP_ST_CONCLUSAO         ESCOLA  IN_TREINEIRO  ...  \
20                  1                3  não_respondeu             1  ...   
21                  1                3  não_respondeu             1  ...   
106                 1                3  não_respondeu             1  ...   
109                 1                3  não_respondeu             1  ...   
159                 1                3  não_respondeu             1  ...   

     NOTA_MT TP_LINGUA  TP_STATUS_REDACAO  COMP1  COMP2  COMP3  COMP4  COMP5  \
20     573.2    Inglês                1.0  140.0  120.0  120.0  100.0   80.0   
21       NaN  Espanhol                NaN    NaN    NaN    NaN    NaN    NaN   
106    532.6    Inglês                1.0  120.0  120.0  120.0  100.0   80.0   
109      NaN  Espanhol                NaN    NaN    NaN    NaN    NaN    NaN   
159      NaN    Inglês                NaN    NaN    NaN    NaN    NaN    NaN   

     NOTA_REDACAO  INTERNET  
20          560.0       Sim  
21            NaN       Não  
106         540.0       Sim  
109           NaN       Sim  
159           NaN       Sim  

[5 rows x 29 columns]

In [42]:
treineiros.to_csv('treineiros_enem_2019_sp.csv', encoding = 'iso-8859-1', index = False)